### PR TITLE
Add DmxBidder and tests

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**
Updated dmx endpoint in dmx.yaml due to https://github.com/prebid/prebid-server/pull/1401

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        for (SeatBid seatBid : bidResponse.getSeatbid()) {
+            for (Bid bid : seatBid.getBid()) {
+                try {
+                    final BidType bidType = getBidType(bid.getImpid(), bidRequest.getImp());
+                    String adm = null;
+                    if (bidType == BidType.video) {
+                        adm = getAdm(bid);
+                    }
+
+                    final Bid updatedBid = bid.toBuilder().adm(adm).build();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:246]:
```
                    final Bid updatedBid = bidType == BidType.video 
                            ? bid.toBuilder().adm(getAdm(bid)).build()
                            : bid;
```

File: src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    public void makeHttpRequestsShouldReturnErrorIfUserIdIsEmpty() {
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .imp(singletonList(Imp.builder()
+                        .ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpDmx.of("tagId", "dmxId", "memberId", "publisherId",
+                                        "sellerId"))))
+                        .build()))
+                .user(User.builder().build())
+                .site(Site.builder().build())
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java:96]: this line is covers NPE which is BUG. Remove from ALL tests

File: src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .imp(singletonList(
+                        Imp.builder()
+                                .id("id")
+                                .banner(Banner.builder()
+                                        .format(singletonList(Format.builder().w(300).h(500).build()))
+                                        .build())
+                                .ext(mapper.valueToTree(ExtPrebid.of(null,
+                                        ExtImpDmx.of("tagId", "dmxId", "memberId", "publisherId",
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java:119]: use builder when there are more then 3 parrameters

File: src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = dmxBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getId, Imp::getTagid, Imp::getExt, Imp :: getSecure)
+                .containsOnly(tuple("id", "dmxId", mapper.valueToTree(ExtPrebid.of(null,
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java:135]: extract long variables for code clearance

File: src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = dmxBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getId, Imp::getTagid, Imp::getExt, Imp :: getSecure)
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/dmx/DmxBidderTest.java:134]: redundant spaces

**Review by DGarbar — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        boolean anyHasId = false;
+        if (request.getApp() != null) {
+            if (StringUtils.isNotBlank(request.getApp().getId())) {
+                anyHasId = true;
+            }
+        }
+
+        final Site site = request.getSite();
+        final Publisher publisher = site.getPublisher();
+        if (site.getPublisher() != null) {
```
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:109]: This check will be redundant

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+        boolean anyHasId = false;
+        if (request.getApp() != null) {
+            if (StringUtils.isNotBlank(request.getApp().getId())) {
+                anyHasId = true;
+            }
+        }
+
+        final Site site = request.getSite();
+        final Publisher publisher = site.getPublisher();
```
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:108]: site.getPublisher() != null this check also site!=null

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:109]: Intellijent Idea propose to siplify and remove this check

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 8]
```diff
@@ -62,39 +62,32 @@ public DmxBidder(String endpointUrl, JacksonMapper mapper) {
...
-                if (i == 0) {
-                    final String publisherId = extImp.getPublisherId();
-                    updatedPublisherId = StringUtils.isBlank(publisherId)
-                            ? publisherId
-                            : extImp.getMemberId();
-                    updatedSellerId = extImp.getSellerId();
+        try {
+            final ExtImpDmx extImp = parseImpExt(imps.get(0));
+            final String publisherId = extImp.getPublisherId();
+            updatedPublisherId = StringUtils.isBlank(publisherId)
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:79]: can be in 1 line

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 9]
```diff
@@ -62,39 +62,32 @@ public DmxBidder(String endpointUrl, JacksonMapper mapper) {
...
+                    ? publisherId
+                    : extImp.getMemberId();
+            updatedSellerId = extImp.getSellerId();
+            for (Imp imp : request.getImp()) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:83]: try need to be inside loop. 
Just create two try catch blocks for extraction of publisher and for imps

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 10]
```diff
@@ -105,33 +98,37 @@ public DmxBidder(String endpointUrl, JacksonMapper mapper) {
...
-                    .site(site.toBuilder().publisher(publisher.toBuilder().id(updatedPublisherId).build()).build())
-                    .build();
-        } else {
-            request.toBuilder()
-                    .site(site.toBuilder().publisher(Publisher.builder().id(updatedPublisherId).build()).build())
-                    .build();
+        if (site != null) {
+            final Publisher publisher = site.getPublisher();
+            if (site.getPublisher() != null) {
+                request.toBuilder()
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:104]: you can do smth like this 
updatedPublisher = site.Publisher == null ? Publisher.builder().id(updatedPublisherId).build() : publisher.toBuilder().id(updatedPublisherId).build()

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 11]
```diff
@@ -210,7 +208,15 @@ private Imp fetchParams(Imp imp, ExtImpDmx extImp) {
                     .build();
         }
 
-        return (updatedImp != null && StringUtils.isBlank(updatedImp.getTagid())) ? imp : updatedImp;
+        return updatedImp != null && StringUtils.isBlank(updatedImp.getTagid()) && StringUtils.isBlank(imp.getTagid())
+                ? null
+                : updatedImp;
+    }
+
+    private void checkIfHasId(boolean anyHasId) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:216]: Not that. 
I mean that we can extract **ALL** related logic to this variable to separate method

File: src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java [hunk 12]
```diff
@@ -210,7 +208,15 @@ private Imp fetchParams(Imp imp, ExtImpDmx extImp) {
                     .build();
         }
 
-        return (updatedImp != null && StringUtils.isBlank(updatedImp.getTagid())) ? imp : updatedImp;
+        return updatedImp != null && StringUtils.isBlank(updatedImp.getTagid()) && StringUtils.isBlank(imp.getTagid())
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/dmx/DmxBidder.java:211]: just return upratedImp

**Review by DGarbar — DISMISSED** [review]

**Review by DGarbar — DISMISSED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**